### PR TITLE
nrf51/temp sens

### PR DESCRIPTION
**Pull request by niklasad1**
Temperatur Sensor for nrf51dk

Simplest possible driver for nrf51dk's temperature sensor
No sampling or other cool stuff... really basic

### CODE REVIEW

File: kernel/src/hil/temp.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+use returncode::ReturnCode;
+
+pub trait TempDriver {
```
Comment (alevy) [inline kernel/src/hil/temp.rs]: Perhaps just `Temperature`

File: kernel/src/hil/temp.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+use returncode::ReturnCode;
```
Comment (alevy) [inline kernel/src/hil/temp.rs]: High level, we can maybe elide this HIL for now. We almost certainly want one to consolidate the interfaces for this controller as well as the TMP006 and SI7021 drivers. However, that might be best saved for a different PR.

File: userland/libtock/temperature.c [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#include "temperature.h"
+
+int temp_init(subscribe_cb callback, void *ud) {
```
Comment (alevy) [inline userland/libtock/temperature.c]: `temperature_init`

File: userland/libtock/temperature.c [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+  return subscribe(DRIVER_TEMP, 0, callback, NULL);
+}
+
+int temp_measure() {
```
Comment (alevy) [inline userland/libtock/temperature.c]: `temperature_measure`

File: userland/libtock/temperature.c [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#include "temperature.h"
```
Comment (alevy) [inline userland/libtock/temperature.c]: I think we only want this in libtock if we keep the temperature HIL.

File: userland/libtock/temperature.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <tock.h>
+
+#define DRIVER_TEMP 36
```
Comment (alevy) [inline userland/libtock/temperature.h]: `#define`s should go after the C++ guards

File: boards/nrf51dk/src/main.rs [hunk 7]
```diff
@@ -133,7 +135,7 @@ pub unsafe fn reset_handler() {
     nrf51::init();
 
     // LEDs
-    let led_pins = static_init!(
+        let led_pins = static_init!(
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: too many indentations

File: boards/nrf51dk/src/main.rs [hunk 8]
```diff
@@ -149,9 +151,9 @@ pub unsafe fn reset_handler() {
     let button_pins = static_init!(
         [&'static nrf51::gpio::GPIOPin; 4],
         [&nrf51::gpio::PORT[BUTTON1_PIN], // 17
-         &nrf51::gpio::PORT[BUTTON2_PIN], // 18
-         &nrf51::gpio::PORT[BUTTON3_PIN], // 19
-         &nrf51::gpio::PORT[BUTTON4_PIN], // 20
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: We actually want these extra spaces so the elements and commas line up

File: boards/nrf51dk/src/main.rs [hunk 9]
```diff
@@ -167,16 +169,16 @@ pub unsafe fn reset_handler() {
...
-         &nrf51::gpio::PORT[3],  //   V
-         &nrf51::gpio::PORT[4],  //
-         &nrf51::gpio::PORT[5],  //
-         &nrf51::gpio::PORT[6],  // -----
-         &nrf51::gpio::PORT[16], //
-         &nrf51::gpio::PORT[15], //
-         &nrf51::gpio::PORT[14], //
-         &nrf51::gpio::PORT[13], //
-         &nrf51::gpio::PORT[12], //
+        &nrf51::gpio::PORT[2],  //   |
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: ditto above

File: boards/nrf51dk/src/main.rs [hunk 10]
```diff
@@ -198,7 +200,7 @@ pub unsafe fn reset_handler() {
                                         115200,
                                         &mut capsules::console::WRITE_BUF,
                                         kernel::Container::create()),
-        224/8);
+                                        224/8);
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: parentheses should close at the same indentation level as they were opened.

File: boards/nrf51dk/src/main.rs [hunk 11]
```diff
@@ -223,9 +225,15 @@ pub unsafe fn reset_handler() {
         TimerDriver<'static, VirtualMuxAlarm<'static, Rtc>>,
         TimerDriver::new(virtual_alarm1,
                          kernel::Container::create()),
-        12);
+                         12);
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: ditto above

File: boards/nrf51dk/src/main.rs [hunk 12]
```diff
@@ -223,9 +225,15 @@ pub unsafe fn reset_handler() {
...
     virtual_alarm1.set_client(timer);
 
+    let temp= static_init!(
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: missing space before `=`

File: boards/nrf51dk/src/main.rs [hunk 13]
```diff
@@ -223,9 +225,15 @@ pub unsafe fn reset_handler() {
...
+        capsules::temp_nrf51dk::Temp<'static, nrf51::temp::Temp>,
+        capsules::temp_nrf51dk::Temp::new(&mut nrf51::temp::TEMP, kernel::Container::create(), &mut capsules::temp_nrf51dk::BUF), 128/8);
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: consistent line length: keep to ~80 chars

**Comment by alevy** [comment]
make sure to run `make fmt` from the top level to fix any rust formatting issues as well.

**Comment by alevy** [comment]
It also looks like there are a bunch of unused import warnings in the temp_nrf51 capsule. Please resolve them by removing unused imports.

File: capsules/src/temp_nrf51dk.rs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        match subscribe_num {
+            0 => {
+                self.apps
+                    .enter(callback.app_id(), |app, _| {
+                        app.callback = Some(callback);
+                        ReturnCode::SUCCESS
+                    })
+                    .unwrap_or_else(|err| match err {
+                        Error::OutOfMemory => ReturnCode::ENOMEM,
+                        Error::AddressOutOfBounds => ReturnCode::EINVAL,
```
Comment (ppannuto) [inline capsules/src/temp_nrf51dk.rs]: @alevy I'm not as sure about this one. This particular idiom is probably copied from lots of other places in the kernel where we list every case. If we switch to a default match arm, then if the return ever changes, anything that deconstructs it will silently point to `EINVAL` - harder to track down refactoring issues.

Comment (alevy) [inline capsules/src/temp_nrf51dk.rs]: OK, I'm convinced by @ppannuto

File: boards/nrf51dk/src/main.rs (hunk 7 above)
Comment (niklasad1) [inline boards/nrf51dk/src/main.rs]: yes, 
"$ make format fmt" should fix all this right?

Comment (ppannuto) [inline boards/nrf51dk/src/main.rs]: Just `make format` (or `make fmt`, they're the same, it's a convenience alias)

File: capsules/src/temp_nrf51dk.rs (hunk 14 above)
Comment (niklasad1) [inline capsules/src/temp_nrf51dk.rs]: I'm not really sure what to do here, shall I keep it as it is? I think we copied this snippet from rng.rs

Comment (ppannuto) [inline capsules/src/temp_nrf51dk.rs]: Yeah, just leave it

File: capsules/src/temp_nrf51dk.rs [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+impl Default for App {
+    fn default() -> App {
+        App { callback: None }
+    }
+}
+
+pub struct Temp<'a, T: TempDriver + 'a> {
+    temp: &'a T,
+    apps: Container<App>,
+    kernel_tx: TakeCell<'static, [u8]>,
```
Comment (niklasad1) [inline capsules/src/temp_nrf51dk.rs]: I don't understand, the struct is used why to delete it?

**Comment by niklasad1** [comment]
well I missed to remove temperature.h and temperature.c in libtock. 
I will fix it later.
Otherwise most stuff should be fixed except "There is a bug in the driver virtualization, and I'm uncertain about including the HIL in this particular PR. It probably makes sense to consolidate this, the tmp006 and isl2071 drivers into a common interface and implement a generic syscall driver/virtualization on top of them.
